python/pyarrow/table.pxi

```diff
@@ -6259,6 +6259,56 @@ def concat_tables(tables, MemoryPool memory_pool=None, str promote_options="none
     return pyarrow_wrap_table(c_result_table)
 
 
+def concat_recordbatches(recordbatches, MemoryPool memory_pool=None):
+    """
+    Concatenate pyarrow.RecordBatch objects.
+
+    All recordbatches must share the same Schema,
```

Comment:
The table version does not require this. That's a missing feature on the C++ side?

Reply:
Yes, the C++ function takes for granted the schema is the same or will error. 
The behavior seems to be consistent with the Rust implementation by the way, which doesn't seem to offer casting options: https://github.com/apache/arrow-rs/blob/d727503e4d52d40d7215a42398b02d58e964ed01/arrow-select/src/concat.rs#L214-L225

Reply:
Not for this PR, but it does look relatively straightforward to port the schema unification from the Table implementation to the batches as well. That might be a nice follow-up issue to create to keep concat_tables / concat_batches consistent

Reply:
Created https://github.com/apache/arrow/issues/44196